src/workers/VolumeLoadWorker.ts

```diff
@@ -0,0 +1,93 @@
+  WorkerResponseKind,
+  WorkerResponsePayload,
+} from "./types";
+import { rebuildImageInfo, rebuildLoadSpec } from "./util";
+
+let cache: VolumeCache | undefined = undefined;
+let queue: RequestQueue | undefined = undefined;
+let subscribableQueue: SubscribableRequestQueue | undefined = undefined;
+let loader: ThreadableVolumeLoader | undefined = undefined;
+let initialized = false;
+let copyOnLoad = false;
+
+type MessageHandler<T extends WorkerMsgType> = (payload: WorkerRequestPayload<T>) => Promise<WorkerResponsePayload<T>>;
+
+const messageHandlers: { [T in WorkerMsgType]: MessageHandler<T> } = {
+  [WorkerMsgType.INIT]: ({ maxCacheSize, maxActiveRequests, maxLowPriorityRequests }) => {
+    if (!initialized) {
+      cache = new VolumeCache(maxCacheSize);
+      queue = new RequestQueue(maxActiveRequests, maxLowPriorityRequests);
+      subscribableQueue = new SubscribableRequestQueue(queue);
+      initialized = true;
+    }
+    return Promise.resolve();
+  },
+
+  [WorkerMsgType.CREATE_LOADER]: async ({ path, options }) => {
+    const pathString = Array.isArray(path) ? path[0] : path;
+    const fileType = options?.fileType || pathToFileType(pathString);
+    copyOnLoad = fileType === VolumeFileFormat.JSON;
+    loader = await createVolumeLoader(path, { ...options, cache, queue: subscribableQueue });
+    return loader !== undefined;
+  },
+
+  [WorkerMsgType.CREATE_VOLUME]: async (loadSpec) => {
+    if (loader === undefined) {
+      throw new Error("No loader created");
+    }
+
+    return await loader.createImageInfo(rebuildLoadSpec(loadSpec));
+  },
+
+  [WorkerMsgType.LOAD_DIMS]: async (loadSpec) => {
+    if (loader === undefined) {
+      throw new Error("No loader created");
+    }
+    return await loader.loadDims(rebuildLoadSpec(loadSpec));
+  },
+
+  [WorkerMsgType.LOAD_VOLUME_DATA]: async ({ imageInfo, loadSpec, loaderId, loadId }) => {
+    if (loader === undefined) {
+      throw new Error("No loader created");
+    }
+
+    return await loader.loadRawChannelData(
+      rebuildImageInfo(imageInfo),
+      rebuildLoadSpec(loadSpec),
+      (channelIndex, data, atlasDims) => {
+        const message: WorkerResponse<WorkerMsgType> = {
+          responseKind: WorkerResponseKind.EVENT,
+          loaderId,
+          loadId,
+          channelIndex,
+          data,
+          atlasDims,
+        };
+        (self as unknown as Worker).postMessage(message, copyOnLoad ? [] : [data.buffer]);
+      }
+    );
+  },
+};
+
+self.onmessage = async <T extends WorkerMsgType>({ data }: MessageEvent<WorkerRequest<T>>) => {
+  const { msgId, type, payload } = data;
+  let message: WorkerResponse<T>;
+
+  try {
+    const response = await messageHandlers[type](payload);
+    message = { responseKind: WorkerResponseKind.SUCCESS, msgId, type, payload: response };
+  } catch (e) {
+    message = { responseKind: WorkerResponseKind.ERROR, msgId, type, payload: (e as Error).message };
+  }
+  self.postMessage(message);
+};
```

Comment:
This is shockingly clean, wow

Reply:
I agree, this whole file is a very elegant way to set up a web worker to handle different kinds of messages!  It looks like tasks that need to transfer data do so in their own postMessage calls, and then those are always followed by a SUCCESS postMessage, right?

Reply:
In the case of the one action with events, the SUCCESS actually comes before all the events. Though that's just a consequence of how the actual function it's calling (`loadRawChannelData`) works: it returns immediately with updates to `ImageInfo` and/or `LoadSpec`, then runs callbacks when data actually arrives. So I suppose you can think of the `postMessage` here as capturing when a function returns, and ones in tasks as capturing callbacks being fired.